File: macos-host/USBDisplayApp/Sources/USBDisplayCore/Protocol.swift
// Protocol.swift — the wire format shared by the macOS host and the Android client.
//
// Two channels run over the same transport (ADB reverse over USB, or TCP over
// Wi-Fi). Both are length-prefixed so a short read can never desynchronise the
// stream — the v1 format used fixed-size records, which meant one dropped byte
// corrupted every frame that followed.
//
//   Video channel  (host -> client): [u32 BE length][Annex-B H.264/HEVC access unit]
//   Input channel  (bidirectional):  [u8 type][u16 BE length][payload]
//
// Keeping the input channel self-describing lets either side skip a message
// type it does not understand, so an older client can talk to a newer host.

import Foundation

public enum WireProtocol {
    /// Bumped whenever a payload layout changes incompatibly.
    public static let version: UInt16 = 2

    public static let defaultVideoPort: UInt16 = 5560
    public static let defaultInputPort: UInt16 = 5561
    /// Port the host's Bonjour service advertises for wireless mode.
    public static let defaultWirelessPort: UInt16 = 5562

    /// Largest video frame we will accept. A 4K keyframe is comfortably under
    /// this; anything larger means the stream is corrupt, not merely big.
    public static let maxFrameBytes = 16 * 1024 * 1024
    /// Largest input message. Input payloads are tiny; the cap is a guard.
    public static let maxInputPayloadBytes = 4096
}

/// Message types on the input channel.
public enum InputMessageType: UInt8, Sendable {
    case touch = 0x01
    case pen = 0x02
    case scroll = 0x03
    case pinch = 0x04
    case hello = 0x10       // client -> host, first message on the channel
    case helloAck = 0x11    // host -> client
    case keepAlive = 0x12   // either direction

    // Wireless pairing. Over USB these are never sent: the cable and adb's own
    // "Allow USB debugging" prompt are the authorisation.
    case pairRequest = 0x20   // client -> host
    case pairResponse = 0x21  // host -> client
    case pairProof = 0x22     // client -> host
    case pairResult = 0x23    // host -> client
}

/// Lifecycle of a contact (finger, pen, or gesture).
public enum TouchPhase: UInt8, Sendable {
    case down = 0
    case move = 1
    case up = 2
    case cancel = 3
    case hover = 4      // pen in range but not touching
    case hoverEnd = 5
}

/// A single finger contact. `pointerId` is stable for the life of the contact,
/// which is what makes two-finger gestures possible at all.
public struct TouchEvent: Equatable, Sendable {
    public var pointerId: UInt8
    public var phase: TouchPhase
    /// Normalised 0...1 across the client's surface.
    public var x: Float
    public var y: Float

    public init(pointerId: UInt8, phase: TouchPhase, x: Float, y: Float) {
        self.pointerId = pointerId
        self.phase = phase
        self.x = x
        self.y = y
    }
}

/// A pen contact. Tilt is carried as altitude/azimuth in radians — the form
/// Android reports natively (AXIS_TILT is a magnitude, AXIS_ORIENTATION a
/// direction) — and converted to macOS's tiltX/tiltY only at injection time.
public struct PenEvent: Equatable, Sendable {
    public var phase: TouchPhase
    public var x: Float
    public var y: Float
    /// 0...1.
    public var pressure: Float
    /// Angle from the surface normal, radians. 0 = perpendicular.
    public var tiltRadians: Float
    /// Direction the pen leans, radians, 0 = towards the top of the display.
    public var orientationRadians: Float
    /// Bit 0 = barrel button, bit 1 = eraser end.
    public var buttons: UInt8

    public init(phase: TouchPhase, x: Float, y: Float, pressure: Float,
                tiltRadians: Float, orientationRadians: Float, buttons: UInt8) {
        self.phase = phase
        self.x = x
        self.y = y
        self.pressure = pressure
        self.tiltRadians = tiltRadians
        self.orientationRadians = orientationRadians
        self.buttons = buttons
    }

    /// macOS wants tilt as two independent -1...1 components. Android gives a
    /// polar pair, so decompose it. tiltX is positive to the right, tiltY
    /// positive towards the top, matching NSEvent's tablet fields.
    public var macOSTilt: (x: Float, y: Float) {
        // sin(tilt) maps 0 (perpendicular) -> 0 and pi/2 (flat) -> 1.
        let magnitude = min(max(sin(tiltRadians), 0), 1)
        return (magnitude * sin(orientationRadians), magnitude * cos(orientationRadians))
    }
}

/// Two-finger scroll, already accumulated on the client into a delta.
public struct ScrollEvent: Equatable, Sendable {
    /// Points, positive = content moves right / up, matching NSEvent.
    public var deltaX: Float
    public var deltaY: Float
    public var phase: TouchPhase

    public init(deltaX: Float, deltaY: Float, phase: TouchPhase) {
        self.deltaX = deltaX
        self.deltaY = deltaY
        self.phase = phase
    }
}

/// Pinch-to-zoom, as an incremental magnification factor.
public struct PinchEvent: Equatable, Sendable {
    /// Incremental change, e.g. 0.02 means "2% larger since the last event".
    public var magnification: Float
    public var phase: TouchPhase

    public init(magnification: Float, phase: TouchPhase) {
        self.magnification = magnification
        self.phase = phase
    }
}

/// What the client tells the host about itself the moment it connects. The
/// host sizes the virtual display from this, which is why the display comes up
/// at the tablet's real resolution and density instead of a hardcoded guess.
public struct ClientHello: Equatable, Sendable {
    public var protocolVersion: UInt16
    public var widthPixels: UInt32
    public var heightPixels: UInt32
    /// Android's DisplayMetrics.densityDpi.
    public var densityDpi: UInt16
    /// Surface rotation in degrees: 0, 90, 180, 270.
    public var rotationDegrees: UInt16
    public var flags: ClientFlags
    /// Human-readable device model, for the Displays pane.
    public var deviceName: String

    public init(protocolVersion: UInt16 = WireProtocol.version,
                widthPixels: UInt32, heightPixels: UInt32, densityDpi: UInt16,
                rotationDegrees: UInt16, flags: ClientFlags, deviceName: String) {
        self.protocolVersion = protocolVersion
        self.widthPixels = widthPixels
        self.heightPixels = heightPixels
        self.densityDpi = densityDpi
        self.rotationDegrees = rotationDegrees
        self.flags = flags
        self.deviceName = deviceName
    }
}

public struct ClientFlags: OptionSet, Equatable, Sendable {
    public let rawValue: UInt8
    public init(rawValue: UInt8) { self.rawValue = rawValue }

    /// Device reports a stylus digitiser (S Pen, or any TOOL_TYPE_STYLUS source).
    public static let hasStylus = ClientFlags(rawValue: 1 << 0)
    /// Device reports pressure resolution better than on/off.
    public static let hasPressure = ClientFlags(rawValue: 1 << 1)
    /// Device reports tilt.
    public static let hasTilt = ClientFlags(rawValue: 1 << 2)
    /// Samsung DeX is currently active — the client warns, the host logs it.
    public static let dexActive = ClientFlags(rawValue: 1 << 3)
    /// User asked for pen-only mode: fingers do not move the pointer.
    public static let penOnly = ClientFlags(rawValue: 1 << 4)
}

/// Host's answer to a hello: what it actually created.
public struct HelloAck: Equatable, Sendable {
    public var protocolVersion: UInt16
    public var accepted: Bool
    public var displayWidth: UInt32
    public var displayHeight: UInt32
    public var message: String

    public init(protocolVersion: UInt16 = WireProtocol.version, accepted: Bool,
                displayWidth: UInt32, displayHeight: UInt32, message: String) {
        self.protocolVersion = protocolVersion
        self.accepted = accepted
        self.displayWidth = displayWidth
        self.displayHeight = displayHeight
        self.message = message
    }
}

/// Client opening a wireless session: who it is, and a fresh nonce.
public struct PairRequest: Equatable, Sendable {
    public var clientKey: Data       // 32 bytes
    public var clientNonce: Data     // 16 bytes
    public var deviceName: String

    public init(clientKey: Data, clientNonce: Data, deviceName: String) {
        self.clientKey = clientKey
        self.clientNonce = clientNonce
        self.deviceName = deviceName
    }
}

/// How the host is treating this client.
public enum PairStatus: UInt8, Sendable {
    /// Never seen before: both sides show the code and a person confirms.
    case needsConfirmation = 0
    /// Already trusted: prove it and carry on, no taps.
    case alreadyTrusted = 1
    /// Refused — wireless is off, or the user declined.
    case rejected = 2
}

public struct PairResponse: Equatable, Sendable {
    public var hostKey: Data         // 32 bytes
    public var hostNonce: Data       // 16 bytes
    public var status: PairStatus
    public var hostName: String

    public init(hostKey: Data, hostNonce: Data, status: PairStatus, hostName: String) {
        self.hostKey = hostKey
        self.hostNonce = hostNonce
        self.status = status
        self.hostName = hostName
    }
}

/// Proof that a returning client still holds the key it paired with.
public struct PairProof: Equatable, Sendable {
    public var proof: Data           // 32 bytes
    public init(proof: Data) { self.proof = proof }
}

public struct PairResult: Equatable, Sendable {
    public var accepted: Bool
    public var message: String
    public init(accepted: Bool, message: String) {
        self.accepted = accepted
        self.message = message
    }
}

/// Anything the input channel can carry, once decoded.
public enum InputMessage: Equatable, Sendable {
    case touch(TouchEvent)
    case pen(PenEvent)
    case scroll(ScrollEvent)
    case pinch(PinchEvent)
    case hello(ClientHello)
    case helloAck(HelloAck)
    case keepAlive
    case pairRequest(PairRequest)
    case pairResponse(PairResponse)
    case pairProof(PairProof)
    case pairResult(PairResult)
    /// A well-formed message of a type we do not know. Carried so the reader
    /// can skip it rather than losing sync.
    case unknown(type: UInt8, payload: Data)
}

File: android-client/app/src/main/java/com/usbtablet/display/Protocol.kt
package com.usbtablet.display

import java.io.ByteArrayOutputStream

/**
 * The wire format, mirroring the macOS host's USBDisplayCore/Protocol.swift.
 *
 * Two channels run over whichever transport is in use (adb reverse over USB,
 * or a plain TCP socket over Wi-Fi):
 *
 *   Video (host -> client): [u32 BE length][Annex-B access unit]
 *   Input (bidirectional):  [u8 type][u16 BE length][payload]
 *
 * Everything is big-endian, which is what DataOutputStream writes natively and
 * what the Swift side reads, so the two implementations stay comparable by eye.
 */
object WireProtocol {
    const val VERSION = 2

    const val DEFAULT_VIDEO_PORT = 5560
    const val DEFAULT_INPUT_PORT = 5561

    const val MAX_FRAME_BYTES = 16 * 1024 * 1024
    const val MAX_INPUT_PAYLOAD_BYTES = 4096
}

object MessageType {
    const val TOUCH: Int = 0x01
    const val PEN: Int = 0x02
    const val SCROLL: Int = 0x03
    const val PINCH: Int = 0x04
    const val HELLO: Int = 0x10
    const val HELLO_ACK: Int = 0x11
    const val KEEP_ALIVE: Int = 0x12
    const val PAIR_REQUEST: Int = 0x20
    const val PAIR_RESPONSE: Int = 0x21
    const val PAIR_PROOF: Int = 0x22
    const val PAIR_RESULT: Int = 0x23
}

/** Lifecycle of a contact. Must match TouchPhase in the Swift core. */
enum class Phase(val code: Int) {
    DOWN(0), MOVE(1), UP(2), CANCEL(3), HOVER(4), HOVER_END(5);

    companion object {
        fun from(code: Int): Phase? = entries.firstOrNull { it.code == code }
    }
}

/** Capability bits the client advertises in its hello. */
object ClientFlags {
    const val HAS_STYLUS = 1 shl 0
    const val HAS_PRESSURE = 1 shl 1
    const val HAS_TILT = 1 shl 2
    const val DEX_ACTIVE = 1 shl 3
    const val PEN_ONLY = 1 shl 4

    // Old Android video stacks can expose hardware H.264 decoders that
    // become unstable with the modern 60 fps / high-bitrate stream.
    const val LEGACY_VIDEO_DECODER = 1 shl 5
}

sealed class OutgoingMessage {
    /** [x] and [y] are normalised 0..1 across the surface. */
    data class Touch(val pointerId: Int, val phase: Phase, val x: Float, val y: Float) :
        OutgoingMessage()

    /**
     * [tiltRadians] is the angle from the surface normal (Android's AXIS_TILT)
     * and [orientationRadians] the direction of lean (AXIS_ORIENTATION). They
     * are deliberately NOT pre-converted to macOS's tiltX/tiltY here: the host
     * does that, so there is exactly one place where the polar-to-cartesian
     * conversion can be wrong.
     */
    data class Pen(
        val phase: Phase, val x: Float, val y: Float, val pressure: Float,
        val tiltRadians: Float, val orientationRadians: Float, val buttons: Int
    ) : OutgoingMessage()

    data class Scroll(val deltaX: Float, val deltaY: Float, val phase: Phase) : OutgoingMessage()
    data class Pinch(val magnification: Float, val phase: Phase) : OutgoingMessage()

    data class Hello(
        val widthPixels: Int, val heightPixels: Int, val densityDpi: Int,
        val rotationDegrees: Int, val flags: Int, val deviceName: String
    ) : OutgoingMessage()

    object KeepAlive : OutgoingMessage()

    data class PairRequest(
        val clientKey: ByteArray, val clientNonce: ByteArray, val deviceName: String
    ) : OutgoingMessage() {
        override fun equals(other: Any?): Boolean =
            other is PairRequest && clientKey.contentEquals(other.clientKey) &&
                clientNonce.contentEquals(other.clientNonce) && deviceName == other.deviceName

        override fun hashCode(): Int =
            (clientKey.contentHashCode() * 31 + clientNonce.contentHashCode()) * 31 +
                deviceName.hashCode()
    }

    data class PairProof(val proof: ByteArray) : OutgoingMessage() {
        override fun equals(other: Any?): Boolean =
            other is PairProof && proof.contentEquals(other.proof)

        override fun hashCode(): Int = proof.contentHashCode()
    }
}

sealed class IncomingMessage {
    data class HelloAck(
        val protocolVersion: Int, val accepted: Boolean,
        val displayWidth: Int, val displayHeight: Int, val message: String
    ) : IncomingMessage()

    data class PairResponse(
        val hostKey: ByteArray, val hostNonce: ByteArray,
        val status: Int, val hostName: String
    ) : IncomingMessage() {
        override fun equals(other: Any?): Boolean =
            other is PairResponse && hostKey.contentEquals(other.hostKey) &&
                hostNonce.contentEquals(other.hostNonce) && status == other.status &&
                hostName == other.hostName

        override fun hashCode(): Int =
            ((hostKey.contentHashCode() * 31 + hostNonce.contentHashCode()) * 31 +
                status) * 31 + hostName.hashCode()
    }

    data class PairResult(val accepted: Boolean, val message: String) : IncomingMessage()
    object KeepAlive : IncomingMessage()
    data class Unknown(val type: Int, val payload: ByteArray) : IncomingMessage() {
        override fun equals(other: Any?): Boolean =
            other is Unknown && type == other.type && payload.contentEquals(other.payload)

        override fun hashCode(): Int = type * 31 + payload.contentHashCode()
    }
}

/** Pairing status codes, matching PairStatus in the Swift core. */
object PairStatus {
    const val NEEDS_CONFIRMATION = 0
    const val ALREADY_TRUSTED = 1
    const val REJECTED = 2
}

// MARK: - Encoding

private class Writer {
    val out = ByteArrayOutputStream()

    fun u8(v: Int) { out.write(v and 0xFF) }
    fun u16(v: Int) { out.write((v ushr 8) and 0xFF); out.write(v and 0xFF) }
    fun u32(v: Int) {
        out.write((v ushr 24) and 0xFF); out.write((v ushr 16) and 0xFF)
        out.write((v ushr 8) and 0xFF); out.write(v and 0xFF)
    }
    fun f32(v: Float) = u32(java.lang.Float.floatToIntBits(v))
    fun raw(bytes: ByteArray) = out.write(bytes)

    /** u8 length prefix then UTF-8, truncated rather than allowed to overflow. */
    fun shortString(s: String) {
        var bytes = s.toByteArray(Charsets.UTF_8)
        if (bytes.size > 255) {
            // Truncate on a character boundary so the result is still valid UTF-8.
            var end = 255
            while (end > 0 && (bytes[end].toInt() and 0xC0) == 0x80) end--
            bytes = bytes.copyOf(end)
        }
        u8(bytes.size)
        raw(bytes)
    }
}

object InputCodec {

    /** Encode a message with its [type][u16 length] header. */
    fun encode(message: OutgoingMessage): ByteArray {
        val body = Writer()
        val type: Int

        when (message) {
            is OutgoingMessage.Touch -> {
                type = MessageType.TOUCH
                body.u8(message.pointerId); body.u8(message.phase.code)
                body.f32(message.x); body.f32(message.y)
            }
            is OutgoingMessage.Pen -> {
                type = MessageType.PEN
                body.u8(message.phase.code)
                body.f32(message.x); body.f32(message.y); body.f32(message.pressure)
                body.f32(message.tiltRadians); body.f32(message.orientationRadians)
                body.u8(message.buttons)
            }
            is OutgoingMessage.Scroll -> {
                type = MessageType.SCROLL
                body.f32(message.deltaX); body.f32(message.deltaY); body.u8(message.phase.code)
            }
            is OutgoingMessage.Pinch -> {
                type = MessageType.PINCH
                body.f32(message.magnification); body.u8(message.phase.code)
            }
            is OutgoingMessage.Hello -> {
                type = MessageType.HELLO
                body.u16(WireProtocol.VERSION)
                body.u32(message.widthPixels); body.u32(message.heightPixels)
                body.u16(message.densityDpi); body.u16(message.rotationDegrees)
                body.u8(message.flags)
                body.shortString(message.deviceName)
            }
            is OutgoingMessage.KeepAlive -> type = MessageType.KEEP_ALIVE
            is OutgoingMessage.PairRequest -> {
                type = MessageType.PAIR_REQUEST
                body.u8(message.clientKey.size); body.raw(message.clientKey)
                body.u8(message.clientNonce.size); body.raw(message.clientNonce)
                body.shortString(message.deviceName)
            }
            is OutgoingMessage.PairProof -> {
                type = MessageType.PAIR_PROOF
                body.u8(message.proof.size); body.raw(message.proof)
            }
        }

        val payload = body.out.toByteArray()
        val framed = Writer()
        framed.u8(type)
        framed.u16(payload.size)
        framed.raw(payload)
        return framed.out.toByteArray()
    }
}

// MARK: - Decoding

class TruncatedMessageException(message: String) : Exception(message)

private class Reader(private val data: ByteArray) {
    private var offset = 0
    val remaining: Int get() = data.size - offset

    fun u8(): Int {
        if (remaining < 1) throw TruncatedMessageException("expected 1 byte, had $remaining")
        return data[offset++].toInt() and 0xFF
    }
    fun u16(): Int = (u8() shl 8) or u8()
    fun u32(): Int = (u8() shl 24) or (u8() shl 16) or (u8() shl 8) or u8()
    fun bytes(count: Int): ByteArray {
        if (remaining < count) throw TruncatedMessageException("expected $count, had $remaining")
        val slice = data.copyOfRange(offset, offset + count)
        offset += count
        return slice
    }
    fun shortString(): String = String(bytes(u8()), Charsets.UTF_8)
}

/**
 * Feeds bytes in, yields whole messages out. A read that lands mid-header is
 * handled the same as one that lands mid-payload, which is the whole point:
 * the previous protocol used fixed-size reads and a single short read
 * desynchronised the channel permanently.
 */
class InputStreamParser {
    private var buffer = ByteArray(0)

    val pendingByteCount: Int get() = buffer.size

    fun append(bytes: ByteArray, count: Int = bytes.size) {
        val combined = ByteArray(buffer.size + count)
        System.arraycopy(buffer, 0, combined, 0, buffer.size)
        System.arraycopy(bytes, 0, combined, buffer.size, count)
        buffer = combined
    }

    /** Next complete message, or null if more bytes are needed. */
    fun next(): IncomingMessage? {
        if (buffer.size < 3) return null

        val type = buffer[0].toInt() and 0xFF
        val length = ((buffer[1].toInt() and 0xFF) shl 8) or (buffer[2].toInt() and 0xFF)
        if (length > WireProtocol.MAX_INPUT_PAYLOAD_BYTES) {
            throw TruncatedMessageException("payload of $length bytes is implausible")
        }
        if (buffer.size < 3 + length) return null

        val payload = buffer.copyOfRange(3, 3 + length)
        buffer = buffer.copyOfRange(3 + length, buffer.size)
        return decode(type, payload)
    }

    private fun decode(type: Int, payload: ByteArray): IncomingMessage {
        val r = Reader(payload)
        return when (type) {
            MessageType.HELLO_ACK -> IncomingMessage.HelloAck(
                protocolVersion = r.u16(),
                accepted = r.u8() != 0,
                displayWidth = r.u32(),
                displayHeight = r.u32(),
                message = r.shortString()
            )
            MessageType.PAIR_RESPONSE -> IncomingMessage.PairResponse(
                hostKey = r.bytes(r.u8()),
                hostNonce = r.bytes(r.u8()),
                status = r.u8(),
                hostName = r.shortString()
            )
            MessageType.PAIR_RESULT -> IncomingMessage.PairResult(
                accepted = r.u8() != 0,
                message = r.shortString()
            )
            MessageType.KEEP_ALIVE -> IncomingMessage.KeepAlive
            // Unknown types are carried, not fatal: that is what lets an older
            // client keep working against a newer host.
            else -> IncomingMessage.Unknown(type, payload)
        }
    }
}

object VideoFraming {
    fun isPlausibleFrameLength(length: Int): Boolean =
        length > 0 && length <= WireProtocol.MAX_FRAME_BYTES
}
